packages/tooltip/test/tooltip-position.test.js

```diff
@@ -0,0 +1,374 @@
+import { css, registerStyles } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
+
+registerStyles(
+  'vaadin-tooltip-overlay',
+  css`
+    [part='overlay'] {
+      width: 50px;
+      border: 1px solid blue;
+    }
+  `,
+);
+
+describe('position', () => {
+  let tooltip, target, overlay;
+
+  beforeEach(() => {
+    tooltip = fixtureSync('<vaadin-tooltip text="tooltip"></vaadin-tooltip>');
+    target = fixtureSync('<div style="width: 100px; height: 100px; margin: 100px; outline: 1px solid red;"></div>');
+    tooltip.target = target;
+    overlay = tooltip.shadowRoot.querySelector('vaadin-tooltip-overlay');
+  });
+
+  async function openAndMeasure() {
+    fire(target, 'mouseenter');
+    await oneEvent(overlay, 'vaadin-overlay-open');
+    const overlayRect = overlay.$.overlay.getBoundingClientRect();
+    const targetRect = target.getBoundingClientRect();
+    return { overlayRect, targetRect };
+  }
+
+  // Overlay above the target (position="top-*")
+  function assertPlacedAbove(overlayRect, targetRect) {
+    expect(overlayRect.bottom).to.be.equal(targetRect.top);
+  }
+
+  // Overlay below the target (position="bottom-*")
+  function assertPlacedBelow(overlayRect, targetRect) {
+    expect(overlayRect.top).to.be.equal(targetRect.bottom);
+  }
+
+  // Overlay before the target (position="start-*")
+  function assertPlacedBefore(overlayRect, targetRect, dir) {
+    const x1 = dir === 'rtl' ? 'left' : 'right';
+    const x2 = dir === 'rtl' ? 'right' : 'left';
+    expect(overlayRect[x1]).to.be.equal(targetRect[x2]);
+  }
+
+  // Overlay after the target (position="end-*")
+  function assertPlacedAfter(overlayRect, targetRect, dir) {
+    const x1 = dir === 'rtl' ? 'right' : 'left';
+    const x2 = dir === 'rtl' ? 'left' : 'right';
+    expect(overlayRect[x1]).to.be.equal(targetRect[x2]);
+  }
+
+  function assertStartAligned(overlayRect, targetRect, dir) {
+    const x = dir === 'rtl' ? 'right' : 'left';
+    expect(overlayRect[x]).to.be.equal(targetRect[x]);
+  }
+
+  function assertEndAligned(overlayRect, targetRect, dir) {
+    const x = dir === 'rtl' ? 'left' : 'right';
+    expect(overlayRect[x]).to.be.equal(targetRect[x]);
+  }
+
+  function assertTopAligned(overlayRect, targetRect) {
+    expect(overlayRect.top).to.be.equal(targetRect.top);
+  }
+
+  function assertBottomAligned(overlayRect, targetRect) {
+    expect(overlayRect.bottom).to.be.equal(targetRect.bottom);
+  }
+
+  function assertCenteredHorizontally(overlayRect, targetRect, dir) {
+    const coord = dir === 'rtl' ? 'right' : 'left';
+    const offset = dir === 'rtl' ? -overlayRect.width / 2 : overlayRect.width / 2;
+    expect(overlayRect[coord]).to.be.equal(targetRect[coord] + offset);
+  }
+
+  function assertCenteredVertically(overlayRect, targetRect) {
+    const offset = targetRect.height / 2 - overlayRect.height / 2;
+    expect(overlayRect.top).to.be.equal(targetRect.top + offset);
+  }
+
+  describe('top-start', () => {
```

Comment:
**idea:** Positioning could theoretically be tested with visual tests which will probably be more concise.

Reply:
Ah, I see that there are such tests already.

Reply:
What are these tests needed for then? Are they for more accuracy?

Reply:
We need unit tests at least to check the actual logic and have code coverage.